File: app/src/main/java/com/xjl/emp3record/MainActivity.java
package com.xjl.emp3record;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import com.xjl.emp3recorder.record_wave.RecordWaveView;

import java.io.File;

public class MainActivity extends AppCompatActivity {

    protected RecordWaveView waveview;

    private static final String cacheDirPath = Environment.getExternalStorageDirectory().getPath() + File.separator + "EMedia";

    public static String mp3CachePath = cacheDirPath + File.separator + "Audios"+File.separator;

    @SuppressLint("NewApi")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        super.setContentView(R.layout.activity_main);
        initView();

        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED &&
                checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermission();
        }


    }

    private void requestPermission() {
        String[] permissions = new String[2];
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.RECORD_AUDIO}, 1);
        }
    }




    private void initView() {
        waveview = (RecordWaveView) findViewById(R.id.waveview);
    }

    public void RealWaveRecord(View view) {
        startActivity(new Intent(this, RealTimeWaveRecordActivity.class));
    }

    public void TouchAudioRecord(View view) {
        startActivity(new Intent(this, TouchAudioRecordActivity.class));
    }

    public void WaveAudioRecord(View view) {
        startActivity(new Intent(this, WaveAudioRecordActivity.class));
    }

}

File: app/src/main/java/com/xjl/emp3/demo/MainActivity.kt
package com.xjl.emp3.demo

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.media.MediaPlayer
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.xjl.emp3record_demo.R
import com.xjl.emp3recorder.logger.Logger
import com.xjl.emp3recorder.record_wave.RecordWaveView
import com.xjl.emp3recorder.utils.MediaPlayerHepler
import com.xjl.emp3recorder.utils.WaveRecordBinder
import java.io.File

/**
 * Created by x33664 on 2019/2/21.
 */
class MainActivity : AppCompatActivity() {

    private val TAG = MainActivity::class.java.simpleName


    private lateinit var mp3CacheDirPath: String

    var waveview: RecordWaveView? = null
    var waveRecordBinder: WaveRecordBinder? = null
    var mediaPlayerHepler: MediaPlayerHepler? = null

    @SuppressLint("NewApi")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        super.setContentView(R.layout.activity_main)
        val permissions = mutableListOf<String>()

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.RECORD_AUDIO)
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q &&
            checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
        ) {
            permissions.add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        }

        if (permissions.isNotEmpty()) {
            requestPermissions(permissions.toTypedArray(), 1)
        }

        mp3CacheDirPath = externalCacheDir?.absolutePath + File.separator + "EMp3Recorder"
        Logger.i("$TAG onCreate mp3CacheDirPath : $mp3CacheDirPath")
        val audioDir = File(mp3CacheDirPath)
        if (!audioDir.exists()) {
            audioDir.mkdirs()
        }

        waveview = findViewById<View>(R.id.waveview) as RecordWaveView
        waveRecordBinder = WaveRecordBinder(waveview!!, mp3CacheDirPath)
        mediaPlayerHepler = MediaPlayerHepler().apply {
            completeListener=object : MediaPlayer.OnCompletionListener{
                override fun onCompletion(p0: MediaPlayer?) {
                  findViewById<TextView>(R.id.player).text = "开始播放"
                }

            }
        }
    }


    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        waveRecordBinder?.onWindowFocusChanged(hasFocus)
    }

    override fun onPause() {
        super.onPause()
        mediaPlayerHepler?.pause()
    }

    var recordaFile: File? = null
    fun recordButton(view: View) {
        waveRecordBinder?.apply {
            if (isRecording) {
                (view as Button).text = "开始录音"
                stop()
            } else {
                (view as Button).text = "停止录音"
                if (ActivityCompat.checkSelfPermission(
                        this@MainActivity,
                        Manifest.permission.RECORD_AUDIO
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    Toast.makeText(this@MainActivity, "没有录音权限", Toast.LENGTH_SHORT).show()
                    return
                }
                recordaFile = start()
            }
        }
    }

    fun player(view: View) {
        mediaPlayerHepler?.let {
            if (it.isPlaying) {
                it.pause()
                (view as Button).text = "开始播放"
            } else {
                if (it.prepared) {
                    it.start()
                } else {
                    it.start(recordaFile?.absolutePath)
                }
                (view as Button).text = "停止播放"
            }
        }
    }
}
